File: autocontext/src/autocontext/harness/scoring/elo.py
"""Elo rating functions — domain-agnostic scoring primitive."""

from __future__ import annotations


def expected_score(player_rating: float, opponent_rating: float) -> float:
    return 1 / (1 + 10 ** ((opponent_rating - player_rating) / 400))


def update_elo(player_rating: float, opponent_rating: float, actual_score: float, k_factor: float = 24.0) -> float:
    expected = expected_score(player_rating, opponent_rating)
    return player_rating + k_factor * (actual_score - expected)

File: ts/src/execution/elo.ts
/**
 * Elo rating functions — domain-agnostic scoring primitive (AC-343 Task 8).
 * Mirrors Python's autocontext/harness/scoring/elo.py.
 */

export function expectedScore(playerRating: number, opponentRating: number): number {
  return 1 / (1 + 10 ** ((opponentRating - playerRating) / 400));
}

export function updateElo(
  playerRating: number,
  opponentRating: number,
  actualScore: number,
  kFactor = 24.0,
): number {
  const expected = expectedScore(playerRating, opponentRating);
  return playerRating + kFactor * (actualScore - expected);
}
